File: src/utils/SignValidator.java
package utils;

import java.util.Stack;

public class SignValidator {
    public boolean isValid(String signos) {

        Stack<Character> stack = new Stack<>();
        for (Character character : signos.toCharArray()) {
            /// character -> (
            /// character -> }
            /// 1 -> comprobar si es de apertura o cierre
            
            if (getIsApertura(character)) {
                // Ingresar a la pila
                stack.push(character);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char ultimo = stack.pop(); // APERTURA
                // Último es del mismo tipo que el character?
                // mismoTipo = isTipoIgual(character, ultimo);
                // ultimo ( && character ) -> TRUE
                // ultimo [ && character ] -> TRUE 
                // ultimo { && character } -> TRUE 
                // ultimo ( && character } -> FALSE 
                if (!isTipoIgual(ultimo, character)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    private boolean getIsApertura(Character character) {
        if (character == '(' || character == '[' || character == '{') {
            return true;
        }
        return false;
    }

    private boolean isTipoIgual(char apertura, char cierre) {
        return (apertura == '(' && cierre == ')') ||
            (apertura == '[' && cierre == ']') ||
            (apertura == '{' && cierre == '}');
    }
}

File: python/sign_validator.py
class SignValidator:

    def is_valid(self, text: str) -> bool:
        stack = []
        pares = {
            ')': '(',
            ']': '[',
            '}': '{'    
        }

        for char in text:
            if char == '(' or char == '{' or char == '[':
                stack.append(char)
            elif char == ')' or char == '}' or char == ']':
                if not stack:
                    return False
                # Aunque pase si no está vacío?
                # comparo el char con el ultimo de la pila
                # si son del mismo tipo saco y avanzo
                # si son de otro return FALSE

                ultimo = stack.pop()

                if pares[char] != ultimo:
                    return False

        return len(stack) == 0